plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,42 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			dnsErrBufOccured := false
+			var perr *dns.Error
+
+			if errors.As(err, &perr) {
+				if errors.Is(err, dns.ErrBuf) {
+					dnsErrBufOccured = true
+				}
+			}
+
+			if proto == "udp" && ((strings.Contains(err.Error(), "overflow")) || dnsErrBufOccured) {
+				newRet := state.Req.Copy()
+
+				// Clear AD bit in case request had set the AD bit. The empty response is not authenticated.
+				newRet.AuthenticatedData = false
+
+				// Clear AA bit in case request had set the AA bit.
+				newRet.Authoritative = false
```

Comment:
Unless we can show that upstream implementations clear the AA and AD bit on a response they truncate themselves, I don't think we should alter the header bits from the original, with the exception of setting the TC bit

Reply:
Thank you. Sure, I have made the suggested changes.